### PR TITLE
Add error handling for interactive #requires

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

Fix for #3803. Restores the tokenizer state more fully to prevent the null dereference causing the original error. Also changed the name of a variable to express that it doesn't actually mean that the PowerShell session is interactive or not.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by rjmholt** [comment]
I've opened [an RFC](https://github.com/PowerShell/PowerShell-RFC/pull/122) on updating the behaviour of `#requires`. Please feel free to comment in the PR.

**Review by daxian-dbw — APPROVED** [review]

**Review by lzybkr — DISMISSED** [review]
Can you add a test?

**Comment by lzybkr** [comment]
@rjmholt - First/Last token are exposed through PowerShell variables `$^` and `$$`, so your test should check the variables.

`Parser.Tests.ps1` seems like a good place for this test.

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 1]
```diff
@@ -915,4 +915,35 @@ foo``u{2195}abc
...
+            @{ script = "#requires -Version 5.0`n10"; firstToken = "10"; lastToken = "10" },
+            @{ script = "Write-Host 'Hello'`n#requires -Version 5.0`n7"; firstToken = "Write-Host"; lastToken = "7" },
+            @{ script = "Write-Host 'Hello'`n#requires -Version 5.0"; firstToken = "Write-Host"; lastToken = "Hello"}
+        )
+
+        It "Correctly resets the first and last tokens in the tokenizer after nested scan in script" -TestCases $testCases {
+            param($script, $firstToken, $lastToken)
+
+            $ps.AddScript($script)
+            $ps.AddScript("(`$^,`$`$)")
```
Comment (lzybkr) [inline test/powershell/Language/Parser/Parser.Tests.ps1]: I'm a little surprised this works, 2 calls to AddScript is I think like piping the first call to the second. That said, it neatly drops any results from the first call.

To be slightly more like the interactive scenario, I think you would call `$ps.Invoke` twice, clearing the commands between calls.

File: ex.ps1 [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Write-Host "Hello"
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply ex.ps1]: Gah, I thought I checked that first. One sec

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 1 above)
Comment (rjmholt) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Yeah, I tried that and it didn't seem to do what I wanted. To be honest, I wasn't entirely certain how I should be using the embedded PowerShell session to recreate the input. But I spent some time playing with it, tried the "script, invoke, tokens, invoke" way and seemed to get `$null` for everything.

**Comment by daxian-dbw** [comment]
@rjmholt Can you please address @lzybkr's comments (_if there is any pending_)?

**Comment by rjmholt** [comment]
@daxian-dbw I think this is all addressed. I just tried `$ps.Invoke()` again, but it's still as I describe in the comment above.

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 3]
```diff
@@ -915,4 +915,35 @@ foo``u{2195}abc
         # Issue #2780
         { ExecuteCommand "`$herestr=@`"`n'`"'`n`"@" } | Should Not Throw
     }
+
+    Context "#requires nested scan tokenizer tests" {
+        BeforeAll {
+            $settings = [System.Management.Automation.PSInvocationSettings]::new()
+            $settings.AddToHistory = $true
+
+            $ps = [powershell]::Create()
```
Comment (daxian-dbw) [inline test/powershell/Language/Parser/Parser.Tests.ps1]: Please have a `AfterAll` block to dispose `$ps`.

File: test/powershell/Language/Scripting/Requires.Tests.ps1 [hunk 4]
```diff
@@ -27,4 +27,14 @@ Describe "Requires tests" -Tags "CI" {
...
-}
\ No newline at end of file
+    Context "Interactive requires" {
+        It "Successfully does nothing when given '#requires' interactively" {
+            $settings = [System.Management.Automation.PSInvocationSettings]::new()
+            $settings.AddToHistory = $true
+
+            $ps = [powershell]::Create()
+            $ps.AddScript("#requires`n10")
+            $ps.Invoke(@(), $settings) | Should -BeExactly 10
```
Comment (daxian-dbw) [inline test/powershell/Language/Scripting/Requires.Tests.ps1]: Please dispose `$ps` after you done with it.
Also, I think it's not necessary to use `PSInvocationSettings` in this test. Just `$ps = [powershell]::Create(); $ps.AddScript("#requires`n10").Invoke() | Should -BeExactly 10` will do.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: test/powershell/Language/Scripting/Requires.Tests.ps1 (hunk 4 above)
Comment (daxian-dbw) [reply test/powershell/Language/Scripting/Requires.Tests.ps1]: I see, we need to hit the $^/$$ setting code. Never mind.

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**